src/esssans/uncertainty.py

```diff
@@ -4,43 +4,63 @@
 
 from typing import Dict, TypeVar, Union, overload
 
+import numpy as np
 import scipp as sc
 
 T = TypeVar("T", bound=Union[sc.Variable, sc.DataArray])
 
 
 @overload
 def broadcast_with_upper_bound_variances(
-    data: sc.Variable, sizes: Dict[str, int]
+    data: sc.Variable, template: sc.Variable
 ) -> sc.Variable:
     pass
 
 
 @overload
 def broadcast_with_upper_bound_variances(
-    data: sc.DataArray, sizes: Dict[str, int]
+    data: sc.DataArray, template: sc.DataArray
 ) -> sc.DataArray:
     pass
 
 
 def broadcast_with_upper_bound_variances(
-    data: Union[sc.Variable, sc.DataArray], sizes: Dict[str, int]
+    data: Union[sc.Variable, sc.DataArray], template: Union[sc.Variable, sc.DataArray]
 ) -> Union[sc.Variable, sc.DataArray]:
-    if _no_variance_broadcast(data, sizes):
+    """
+    Upper-bound estimate for errors from broadcasting a dense array to match the sizes
+    of a given template array.
+
+    We first count the number of non-masked elements in the template. To get the
```

Comment:
I think this approach only works if we are sure that we will subsequently perform a reduction operation along the dimensions that contain masks.

If we never do a reduction, but just compare say a normalization term with `pixel` and `wavelength` dimensions, if some of the pixels were masked, the variances of the normalization term will be different everywhere, compared to a case where no pixels are masked.

This doesn't seem to make much sense conceptually?

Reply:
Not sure why they would be "different everywhere"?

The upper-bound approach actually ensures that the errors are consistent (and upper bound) for *any* subsequent operation, so even if we never normalize and reduce I think it makes sense?

Reply:
> Now say half of the template is masked, we are now only multiplying the variances by 3. If everything stops here, and that is the final result I want to look at, it seems inconsistent?

I think it is consistent. We return a result with no correlations, they have been "absorbed" into the error bars.